transport/transport.go

```diff
@@ -228,22 +223,19 @@ func (t *WebSocketTransport) Heartbeat(ctx context.Context, interval, timeout in
 // a ClosedError or a ConnectionError if unable to receive a message. Receive
 // blocks until the connection has a message ready or a timeout is reached.
 func (t *WebSocketTransport) Receive() (*Message, error) {
-	t.mutex.RLock()
-	if t.closed {
-		t.mutex.RUnlock()
+	t.readMu.Lock()
+	defer t.readMu.Unlock()
+
+	if t.Closed() {
```

Comment:
There is still the infinitissimal risk that the connection will be closed between the time t.Closed() is called and the time ReadMessage() is called. However since ReadMessage() will return the error anyways I don't believe this is an issue anyways.

Reply:
Yes, I believe this will actually occur quite frequently in practice. So we have an instance of optimistic locking for the common case, with the less optimal case being that we actually try the connection, which fails predictably, like you say. I _think_ it's OK...

Reply:
Agreed. But you had me think... a long Send() - whether it's a large message or slow network - can block the reader(s) since they still require a lock to `t.writeMu`. I'm not sure in practice how impactful this will be.